### DISCUSSION TITLE
Inférence de bornes pour les règles numériques

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#391] [Ideas]
Pour certaines valeurs numériques telles que l'âge du salarié ou bien l'effectif de l'entreprise, nous n'avons pas besoin de connaître la valeur exacte dans les simulateurs mais uniquement dans quel intervalle elle se situe.

En effet à chaque fois que l'on référence `entreprise . effectif` dans la base de règle, c'est pour faire une comparaison `entreprise . effectif > 50 salariés` et jamais `entreprise . effectif × quelque chose` (sauf sur la taxe sur les salaires mais qui ne s'applique pas dans la plupart des cas et dont le calcul est de toute façon plus qu'approximatif, bref.).

L'idée serait donc de récupérer l'ensemble de ces paliers utilisés dans les règles actives pour générer automatiquement une question du type :

> Quel est l'effectif de l'entreprise ?
> - 10 salariés 
> - entre 11 et 50 salariés
> - 51 et plus

C'est déjà un peu ce que l'on fait manuellement aujourd'hui, mais l'idée serait d'automatiser ce type de question ainsi que de filtrer les paliers inutiles (exemple : pour un stagiaire le seuil de 50 salariés ne change rien on peut donc discriminer entre les 2 premiers paliers seulement; exemple 2 : paliers différents pour l'outre-mer). 

La situation est comparable pour la question sur l'âge du salarié, et est bien moins bien géré aujourd'hui. Il existe différentes aides à l'embauche :
- aide pour les moins de 26 ans en CDI ou CDD
- aide pour les moins de 30 ans en apprentissage ou professionnalisation
- aide pour les plus de 45 ans en contrat de professionnalisation

Ainsi si on sait qu'on est dans le cadre d'un contrat de professionnalisation, il faut demander :
> Quel est l'âge du salarié ?
> - 30 ans et moins
> - entre 31 et 44 ans
> - 45 ans et plus

Et si c'est un CDI,
> Le salarié a-t-il 26 ans ou moins ?
> - oui
> - non

De même dans les autres cas de figure avec les bonnes bornes.

Pour info aujourd'hui on gère ça avec une série de question sur l'âge qui "rendent non applicable" les autres questions du même type, mais c'est assez fragile et ça n'est passable à l'échelle.

**Comment by denismerigoux** [comment]
Vous pourriez faire ça avec une petite passe d'analyse statique. Pour chaque variable `X`, vous regardez toutes les expressions où `X` est utilisée. Si toutes ces expressions sont de la forme `X < n` ou `X > n`, alors vous pouvez récuprérer la liste des `n` et ça vous fait vos bornes.

**Comment by mquandalle** [comment]
Avec quand même une subtilité qui est qu'il faut invalider la réponse quand les paliers changent suite à une modification de la situation (exemple finalement c'est plus un CDI mais un apprenti).

On pourrait même avoir une invalidation intelligente, du type si j'ai répondu "moins de 26 ans" et que la nouvelle question est "moins de 30 ans ?", je peux automatiquement répondre à la nouvelle question sans l'exposer.

Pour info on a aussi un cas d'usage dans le simulateur d'impôt sur les sociétés betagouv/mon-entreprise#1230. Il faut poser une question "le chiffre d'affaire est-il inférieur à 2,5 millions d'euros ?" où 2,5 est proratisé en fonction de la durée de l'exercice. Donc si après avoir répondu à cette question, la personne précise que l'exercice a durée 18 mois au lieu de 12 mois utilisés par défaut, il faut recalculer le nouveau palier mais pas besoin de reposer la question (si on est inférieur à 2,5 on est aussi inférieur à 2,5×150%), en revanche si l'exercice a duré 6 mois il faut reposer la question en "demandant le CA est-il inférieur à 2,5*50% = 1,25 million d'euros ?"

**Comment by mquandalle** [comment]
J'ai eu le cas d'usage sur MesAidesVélo. La plupart des aides dépendent du revenu fiscal de référence, mais on ne veut pas demander cette information avec un champ `input` dans l'interface car on n'a pas besoin de la valeur exacte mais seulement de savoir dans quelle « catégorie » de revenu se trouve la personne.

De plus les aides utilisent des seuils différents, on ne peut donc pas coder cette logique “en dur”.

L'approche consiste à faire une passe d'analyse statique sur les règles pour identifier toutes les expressions de type comparaison dont l'une des opérandes est la règle recherchée, par exemple dans `applicable si: revenu > 3000€` on récupère [3000].

On peut le faire soit globalement sur l'ensemble des règles, soit seulement sur une sous partie nous intéresse. C'est ce que j'ai choisi de faire sur MesAidesVélo, mais l'approche globale fonctionne aussi. Sur MesAidesVélo on trouvait une quinzaine de seuils de revenus différents sur l'ensemble des règles.

Ensuite dans un cas particulier, tous les seuils ne sont pas forcément “actifs”, car la condition peut se trouver dans une sous-partie de l'arbre qui n'est pas utilisée dans le calcul courant.

Pour détecter les seuils actifs l'approche consiste à évaluer l'objectif, ici le montant des aides, en-dessous et au-dessus du seuil.

```js
let isActiveThreshold = evaluate('aides', {...situation, revenu: threshold - 1}).nodeValue !== evaluate('aides', {...situation, revenu: threshold + 1}).nodeValue
```

Il y a plusieurs optimisations possibles sur cet algo naïf. Déjà vu que l'on évalue tous les seuils on peut réutiliser la borne supérieure du seuil précédent comme borne inférieure du seuil courant.

Ensuite si on a une large liste de seuils à évaluer, et vu que notre fonction est monotone, on peut procéder par dichotomie avec un algorithme du type :

```js
function evaluateThresholds(list) {
  if (list.length <= 2)
    return list.map(evaluate)
  
  let start = evaluate(list[0])
  let end = evaluate(list.at(-1))
  if (start === end) 
    return Array(list.length).fill(start)

  let pivot = Math.ceil(list.length / 2)
  return [
    ...evaluateThresholds(list.slice(0, pivot)),
    ...evaluateThresholds(list.slice(pivot, list.length))
  ]
}
```

L'idée est que si sur un intervalle, le premier et le dernier éléments ont la même valeur, alors c'est le cas aussi de toutes les valeurs intermédiaires. On peut ensuite associer les valeurs évaluées avec les seuils entrés et éliminer les seuils inactifs avec un `reduce`.

Sur MesAidesVélo je n'ai pas implémenté cette recherche par dichotomie car en restreignant l'analyse statique initiale aux seules règles affichées à l'écran, la liste des règles à évaluer est de faible longueur (moins de 5 éléments). Par ailleurs dans mon cas d'usage le calcul de l'aide est très peu coûteux.

Un exemple de l'UI avec la détection des seuils : https://mesaidesvelo.fr/ville/albi?velo=électrique

Il est possible que l'on puisse ré-intégrer certaines parties de cette logique dans `publicodes` pour faciliter l'implémentation de ce type de fonctionnalité par les utilisateurs. Mais il faudrait sans doute d'autres cas d'usage implémentés afin d'identifier les bonnes abstractions à mettre dans la librairie.
